dev/devicelab/lib/tasks/perf_tests.dart

```diff
@@ -320,13 +337,148 @@ class PerfTest {
+      _flutterPath = path.join(flutterDirectory.path, 'bin', 'flutter');
+
+      // Prepare the SkSL by running the driver test.
+      await _generateSkSL();
+
+      // Build the app with SkSL artifacts and run that app
+      final String observatoryUri = await _buildAndRun();
+
+      // Attach to the running app and run the final driver test to get metrics.
+      final TaskResult result = await internalRun(
+        existingApp: observatoryUri,
+      );
+
+      _runProcess.kill();
+      await _runProcess.exitCode;
+
+      return result;
+    });
+  }
+
+  Future<void> _generateSkSL() async {
+    // First, clear all old sksl.json file so they won't affect the test.
+    for (final FileSystemEntity file in Directory(testDirectory).listSync()) {
+      if (file.path.endsWith('.sksl.json')) {
+        file.deleteSync();
+      }
+    }
+
+    await super.internalRun(keepRunning: true, cacheSkSL: true);
+    final Process process = await startProcess(
+      _flutterPath,
+      <String>[
+        'attach',
+        '-d', _device.deviceId,
+      ],
+    );
+    final Stream<List<int>> broadcastOut = process.stdout.asBroadcastStream();
+    _forwardStream(broadcastOut, 'attach stdout');
+    _forwardStream(process.stderr, 'attach stderr');
+
+    final Completer<bool> attachReady = Completer<bool>();
+    final Completer<bool> skslWritten = Completer<bool>();
+
+    _transform(broadcastOut).listen((String line) {
+      if (!attachReady.isCompleted && _kObservatoryReadyRegExp.hasMatch(line)) {
+        attachReady.complete(true);
+      }
+      if (!skslWritten.isCompleted && RegExp('Wrote SkSL data to').hasMatch(line)) {
+        skslWritten.complete(true);
+      }
+    });
+
+    await attachReady.future;
+    process.stdin.write('M');
+    await skslWritten.future;
+    process.stdin.write('q');
+    await process.exitCode;
+  }
+
+  // Return the Observatory URI.
+  Future<String> _buildAndRun() async {
+    await flutter('build', options: <String>[
+      // TODO(liyuqian): also supports iOS once https://github.com/flutter/flutter/issues/53115 is fully closed.
+      'apk',
+      '--profile',
+      '--bundle-sksl-path', '$testDirectory/flutter_01.sksl.json',
+      '-t', testTarget,
+    ]);
+
+    _runProcess = await startProcess(
+      _flutterPath,
+      <String>[
+        'run',
+        '--verbose',
+        '--profile',
+        '-d', _device.deviceId,
+        '-t', testTarget,
+        '--use-application-binary', '$testDirectory/build/app/outputs/flutter-apk/app-profile.apk',
+        // Getting Observatory URI from --vmservice-out-file is flaky as we're
+        // not sure when the file is written. Hence we
+      ],
+    );
+
+    final Stream<List<int>> broadcastOut = _runProcess.stdout.asBroadcastStream();
+    _forwardStream(broadcastOut, 'run stdout');
+    _forwardStream(_runProcess.stderr, 'run stderr');
+
+    // Getting Observatory URI from --vmservice-out-file is flaky as we're
```

Comment:
You should delete the file before the test starts and then poll for the existence, that avoid scraping stdout which I 100% guarantee is less reliable. See https://github.com/flutter/flutter/blob/master/dev/devicelab/lib/tasks/track_widget_creation_enabled_task.dart#L118